File: mail/src/main/java/com/example/ApplicationEventListener.java
package com.example;

import java.util.Map;

import org.springframework.boot.context.event.ApplicationReadyEvent;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

@Component
public class ApplicationEventListener {
    private final EmailSender emailSender;

    public ApplicationEventListener(EmailSender emailSender) {
        this.emailSender = emailSender;
    }

    @EventListener(ApplicationReadyEvent.class)
    public void readyEvent() {
        emailSender.send("to1@example.com", Map.of("name", "user1"));
        emailSender.send("to2@example.com", Map.of("name", "user2"));
    }
}

File: kotlin/spring-data/spring-data-r2dbc/src/main/kotlin/com/example/ApplicationEventListener.kt
package com.example

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.boot.context.event.ApplicationReadyEvent
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Component
import org.springframework.transaction.reactive.TransactionalOperator
import reactor.core.publisher.Mono

@Component
class ApplicationEventListener(
    private val operator: TransactionalOperator,
    private val userRepository: UserRepository
) {
    companion object {
        private val log: Logger = LoggerFactory.getLogger(this::class.java)
    }

    @EventListener(ApplicationReadyEvent::class)
    fun readyEvent() {
        val user1: Mono<User> =
            userRepository.findById(1)
                .map { user -> User(user.id, user.name + "!") }
                .flatMap { userRepository.save(it) }
        operator.transactional(user1)
            .subscribe { log.info("$it") }
    }
}
